lsat/token.go

```diff
@@ -0,0 +1,188 @@
+package lsat
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"gopkg.in/macaroon-bakery.v2/bakery/checkers"
+	"gopkg.in/macaroon.v2"
+)
+
+// Token is the main type to store an LSAT token in.
+type Token struct {
+	// BaseMac is the binary serialized base macaroon in its original
+	// form as baked by the authentication server. No client side caveats
+	// have been added to it yet.
+	BaseMac []byte
+
+	// PaymentHash is the hash of the LSAT invoice that needs to be paid.
+	// Knowing the preimage to this hash is seen as proof of payment by the
+	// authentication server.
+	PaymentHash lntypes.Hash
+
+	// Preimage is the proof of payment indicating that the token has been
+	// paid for if set.
+	Preimage lntypes.Preimage
+
+	// AmountPaid is the total amount in msat that the user paid to get the
+	// token. This does not include routing fees.
+	AmountPaid lnwire.MilliSatoshi
+
+	// RoutingFeePaid is the total amount in msat that the user paid in
+	// routing fee to get the token.
+	RoutingFeePaid lnwire.MilliSatoshi
+
+	// TimeCreated is the moment when this token was created.
+	TimeCreated time.Time
+}
+
+// NewToken creates a new token from the given base macaroon and payment
+// information.
+func NewToken(baseMac []byte, paymentHash *[32]byte, preimage lntypes.Preimage,
+	amountPaid, routingFeePaid lnwire.MilliSatoshi) (*Token, error) {
+
+	token, err := tokenFromChallenge(baseMac, paymentHash)
+	if err != nil {
+		return nil, err
+	}
+	token.Preimage = preimage
+	token.AmountPaid = amountPaid
+	token.RoutingFeePaid = routingFeePaid
+	return token, nil
+}
+
+// tokenFromChallenge parses the parts that are present in the challenge part
+// of the LSAT auth protocol which is the macaroon and the payment hash.
+func tokenFromChallenge(baseMac []byte, paymentHash *[32]byte) (*Token, error) {
+	// First, validate that the macaroon is valid and can be unmarshaled.
+	mac := &macaroon.Macaroon{}
+	err := mac.UnmarshalBinary(baseMac)
```

Comment:
Nit: store the serialized macaroon instead?

Reply:
Yeah, makes sense.